test/Kestrel.Core.Tests/Http2ConnectionTests.cs

```diff
@@ -2886,6 +2954,158 @@ public async Task ConnectionResetNotLoggedWithNoActiveStreams()
             Assert.DoesNotContain(_logger.Messages, m => m.Exception is ConnectionResetException);
         }
 
+        [Fact]
+        public async Task OnInputOrOutputCompletedSendsFinalGOAWAY()
+        {
+            await InitializeConnectionAsync(_noopApplication);
+
+            _connection.OnInputOrOutputCompleted();
+            Assert.True(_connection.StateTransitionEvents[Http2Connection.ConnectionState.Closed].Wait(5000));
+
+            VerifyGoAway(await ReceiveFrameAsync(), 0, Http2ErrorCode.NO_ERROR);
+        }
+
+        [Fact]
+        public async Task AbortSendsFinalGOAWAY()
+        {
+            await InitializeConnectionAsync(_noopApplication);
+
+            _connection.Abort(new ConnectionAbortedException());
+            Assert.True(_connection.StateTransitionEvents[Http2Connection.ConnectionState.Closed].Wait(5000));
+
+            VerifyGoAway(await ReceiveFrameAsync(), 0, Http2ErrorCode.INTERNAL_ERROR);
+        }
+
+        [Fact]
+        public async Task CompletionSendsFinalGOAWAY()
+        {
+            await InitializeConnectionAsync(_noopApplication);
+
+            // Completes ProcessRequestsAsync
+            _pair.Application.Output.Complete();
+            Assert.True(_connection.StateTransitionEvents[Http2Connection.ConnectionState.Closed].Wait(5000));
+
+            VerifyGoAway(await ReceiveFrameAsync(), 0, Http2ErrorCode.NO_ERROR);
+        }
+
+        [Fact]
+        public async Task GracefulGOAWAYSentByStopProcessingNextRequest()
+        {
+            await InitializeConnectionAsync(_waitForAbortApplication);
+
+            await StartStreamAsync(1, _browserRequestHeaders, endStream: true);
+
+            Assert.Equal(Http2Connection.ConnectionState.Open, _connection.State);
+            Assert.Equal(1, _connection.HighestOpenedStreamId);
```

Comment:
Your's seem to be the only Http2ConnectionTests facts that access _connection directly. I don't think it's good to verify internal state. If it's important that certain state transitions happen before continuing the test to avoid races, I suggest hooking the logger for this purpose.